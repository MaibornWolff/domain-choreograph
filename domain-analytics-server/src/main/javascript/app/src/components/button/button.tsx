import * as React from 'react';
import { styled } from '~utils/styled';

export interface ButtonProps {
  lightColor?: boolean;
}

export const Button = styled.button<ButtonProps>`
  border: none;
  padding: .5rem;
  background: none;
  font-size: 1rem;
  cursor: pointer;
  border-radius: ${({theme}) => theme.dimensions.borderRadius.normal};
  color: ${({ lightColor, theme }) =>
  lightColor
    ? theme.colors.text.light.primary
    : theme.colors.text.dark.primary
  };

  &:hover, &:focus {
    outline: none;
    background: rgba(0, 0, 0, .15);
  }
`;
