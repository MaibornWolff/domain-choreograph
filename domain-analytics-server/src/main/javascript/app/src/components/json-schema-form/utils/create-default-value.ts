import { JSONSchema4, JSONSchema4TypeName } from 'json-schema';

const defaultValues: Record<Exclude<JSONSchema4TypeName, 'object'>, any> = {
  string: '',
  number: undefined,
  integer: undefined,
  boolean: false,
  array: [],
  'null': null,
  any: ''
};

export function createDefaultValue(schema: JSONSchema4): object {
  return createDefaultValueForObject(schema);
}

function createDefaultValueForObject(schema: JSONSchema4): Promise<object> {
  if (schema.type !== 'object') {
    throw new Error('Expect schema to be an object');
  }
  let defaultValue: any = {};
  if (schema.properties) {
    Object.entries(schema.properties).forEach(([key, property]) => {
      const type = property.type || 'any';
      if (type instanceof Array) {
        throw new Error('Array types are no supported');
      } else if (type === 'object') {
        defaultValue[key] = createDefaultValueForObject(property);
      } else {
        defaultValue[key] = defaultValues[type];
      }
    });
  }
  return defaultValue;
}
