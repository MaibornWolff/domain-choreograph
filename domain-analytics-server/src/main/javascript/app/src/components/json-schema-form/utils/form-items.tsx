import { styled } from '~utils/styled';

export const FormItems = styled('div')`
    display: flex;
    flex-direction: column;

    & > * {
      margin-bottom: .5rem;
    }
`;
