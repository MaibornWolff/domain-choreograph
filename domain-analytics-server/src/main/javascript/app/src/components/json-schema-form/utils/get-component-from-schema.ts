import { FieldComponent } from 'clean-forms';
import { JSONSchema4, JSONSchema4TypeName } from 'json-schema';
import { DatePickerMsField } from '~components/json-schema-form/form-fields/date-picker-ms-field';
import { NumberTextField, TextField } from '~components/json-schema-form/form-fields/text-field';
import { JsonSchemaFormFieldProps } from '~components/json-schema-form/models/json-schema-form-field-props';

export interface JsonSchemaComponentMap {
  matcher?: { type: JSONSchema4TypeName, format?: string };
  field: FieldComponent<any, JsonSchemaFormFieldProps>;
}

const defaultMapping: JsonSchemaComponentMap[] = [
  { matcher: { type: 'integer', format: 'utc-millisec' }, field: DatePickerMsField },
  { matcher: { type: 'integer' }, field: NumberTextField },
  { matcher: { type: 'number' }, field: NumberTextField },
  { field: TextField }
];

export function getComponentFromSchema(
  schema: JSONSchema4,
  mapping: JsonSchemaComponentMap[] = defaultMapping
): FieldComponent<any, JsonSchemaFormFieldProps> {
  for (const map of mapping) {
    if (map.matcher == null) {
      return map.field;
    }
    const typeIsMatching = map.matcher.type === schema.type;
    const formatIsMatching = map.matcher.format == null || map.matcher.format === schema.format;
    if (typeIsMatching && formatIsMatching) {
      return map.field;
    }
  }
  throw new Error('Couldn\'t match the schema');
}
