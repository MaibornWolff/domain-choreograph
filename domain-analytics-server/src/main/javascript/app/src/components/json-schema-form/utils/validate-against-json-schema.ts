import { ValidationFunction } from 'clean-forms';
import { JSONSchema4 } from 'json-schema';
import { Validator } from 'jsonschema';


export function validateAgainstJsonSchema(schema: JSONSchema4): ValidationFunction {
  const validator = new Validator();
  return ({ value }) => {
    const result = validator.validate(value, schema as any);
    const errors = result.errors.map(error => {
      let message = error.message;
      let property = error.property
        .replace(/^instance\.?/, '');
      let schema = error.schema as JSONSchema4;
      if (error.name === 'required') {
        message = `is required`;
        schema = schema.properties![error.argument];
        property = joinProperties(property, error.argument);
      }

      const title = schema.title || '';
      message = `${title} ${message}`;

      return [ property, message ] as [string, string];
    });
    return errors;
  }
}

function joinProperties(path1: string, path2: string): string {
  if (path1 === '') {
    return path2;
  }
  if (path2 === '') {
    return path1;
  }
  return `${path1}.${path2}`;
}
