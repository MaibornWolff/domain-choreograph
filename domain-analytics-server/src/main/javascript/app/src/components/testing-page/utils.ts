export function escapeJavaClass(javaClass: string) {
  return javaClass.replace(/\./g, '-');
}

export function reverseJavaClassEscape(javaClass: string) {
  return javaClass.replace(/-/g, '.');
}
