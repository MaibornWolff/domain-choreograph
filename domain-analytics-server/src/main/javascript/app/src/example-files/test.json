{
  "nodes": {
    "A": {
      "id": "A",
      "name": "A",
      "scope": "root"
    }
  },
  "dependencies": [],
  "scopes": {
    "root": {
      "id": "root",
      "executionContext": "Application",
      "nodes": [
        "A"
      ]
    }
  },
  "executionContexts": {
    "Application": {
      "id": "Application",
      "scopes": [
        "root"
      ]
    }
  }
}
