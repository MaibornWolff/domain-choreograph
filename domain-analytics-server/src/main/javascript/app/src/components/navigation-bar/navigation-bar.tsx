import * as React from 'react';
import { Link } from 'react-router-dom';
import { Button } from '~components/button/button';
import { ThemeKey } from '~styles/theme';
import { ChoreoExecutionContext, ChoreoGraph, ChoreoScope } from '~types/choreo-graph';
import { styled } from '~utils/styled';

const StyledNav = styled('nav')`
  display: flex;
  align-items: center;
  justify-content: space-between;
  height: 3.5rem;
  padding: 0 .8rem;
  background: ${({ theme }) => theme.colors.primary[0]};
  color: ${({ theme }) => theme.colors.text.light.primary};
`;

const NavItems = styled('div')`
  display: flex;
`;

const Ul = styled('ul')`
  display: flex;
  list-style-type: none;
  margin: 0;
  padding: 0;
`;

const StyledLabel = styled('span')`
  padding: .5rem;
  cursor: default;
`;

const StyledLink = styled(Link)`
  color: inherit;
  text-decoration: none;
  padding: .5rem;
`;

const Separator = styled('span')`
  color: ${({ theme }) => theme.colors.text.light.secondary}
`;

export interface NavigationBarProps {
  graph: ChoreoGraph | null;
  executionContext: string;
  scope?: string;
  currentThemeKey?: ThemeKey;
  setThemeKey?: (theme: ThemeKey) => void;
}

export interface NavigationBarState {
}

export class NavigationBar extends React.Component<NavigationBarProps, NavigationBarState> {
  public render() {
    const { graph, executionContext, scope, setThemeKey, currentThemeKey } = this.props;
    if (!graph) {
      return null;
    }

    let startingPoint: NavigationNode;
    if (scope) {
      startingPoint = { type: 'scope', item: graph.scopes[scope] };
    } else {
      startingPoint = { type: 'context', item: graph.executionContexts[executionContext] };
    }
    const navigationPath = getNavigationPath(startingPoint, graph);

    return (
      <StyledNav>
        <NavItems>
          <Ul>
            {
              navigationPath.map((node, i) => {
                const isLast = i === navigationPath.length - 1;

                let link: JSX.Element;
                if (isLast) {
                  link = <StyledLabel>{node.item.id}</StyledLabel>;
                } else {
                  link = node.type === 'context'
                    ? <StyledLink to={`/${node.item.id}`}>{node.item.id}</StyledLink>
                    : <StyledLink to={`/${navigationPath[i - 1].item.id}/${node.item.id}`}>{node.item.id}</StyledLink>;
                }
                return <li key={node.type + node.item.id}>
                  {link}
                  {!isLast && <Separator>></Separator>}
                </li>;
              })
            }
          </Ul>
        </NavItems>
        <NavItems>
          {setThemeKey && currentThemeKey && <Button lightColor={true} onClick={this.switchTheme}>Change Theme</Button>}
        </NavItems>
      </StyledNav>
    );
  }

  private switchTheme = () => {
    const { setThemeKey, currentThemeKey } = this.props;
    if (!setThemeKey || !currentThemeKey) {
      return;
    }

    if (currentThemeKey === 'dark') {
      setThemeKey('light');
    } else {
      setThemeKey('dark');
    }
  }
}

type NavigationNode = {
  type: 'context';
  item: ChoreoExecutionContext;
} | {
  type: 'scope';
  item: ChoreoScope;
};

function getNavigationPath(startingPoint: NavigationNode | undefined, graph: ChoreoGraph): NavigationNode[] {
  console.log(startingPoint, graph);
  let result: NavigationNode[] = [];

  if (!startingPoint) {
    return result;
  }

  let parentNode: NavigationNode;
  if (startingPoint.type === 'scope') {

    const parentContext = graph.executionContexts[startingPoint.item.executionContext];
    parentNode = { type: 'context', item: parentContext };
    if (parentContext.scopes.length > 1) {
      result = [startingPoint];
    }

  } else if (startingPoint.type === 'context') {

    const node = graph.nodes[startingPoint.item.id];
    result = [startingPoint];
    if (!node) {
      return result;
    }
    const parentScopeId = node.scope;
    const parentScope = graph.scopes[parentScopeId];
    console.log('PARENT SCOPE', parentScopeId, parentScope);
    parentNode = { type: 'scope', item: parentScope };

  } else {
    throw new Error(`Invalid navigation node type.`);
  }

  return [...getNavigationPath(parentNode, graph), ...result];
}
