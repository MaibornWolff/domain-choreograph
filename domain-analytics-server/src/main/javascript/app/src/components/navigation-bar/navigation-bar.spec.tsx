import { shallow } from 'enzyme';
import * as React from 'react';
import { Button } from '~components/button/button';
import { graphMock } from '~types/cheoreo-graph.mock';
import { shallowWithProviders } from '~utils/test.utils';
import { NavigationBar } from './navigation-bar';

describe('NavigationBar', () => {
  it('should render without graph', () => {
    expect(shallowWithProviders(<NavigationBar graph={null} executionContext={'root'}/>)).toMatchSnapshot();
  });

  it('should render with graph and without scope', () => {
    expect(shallowWithProviders(<NavigationBar graph={graphMock} executionContext={'root'}/>)).toMatchSnapshot();
  });

  it('should render with graph and scope', () => {
    expect(shallowWithProviders(<NavigationBar graph={graphMock} executionContext={'root'} scope={'root'}/>)).toMatchSnapshot();
  });

  it('should switch theme to dark on button click', () => {
    const setTheme = jest.fn();
    const element = shallow(<NavigationBar
      setThemeKey={setTheme}
      currentThemeKey={'light'}
      graph={graphMock}
      executionContext={'root'}
      scope={'root'}
    />);
    const button = element.find(Button);

    button.simulate('click');
    expect(setTheme).toBeCalledWith('dark');
  });

  it('should switch theme to default on button click', () => {
    const setTheme = jest.fn();
    const element = shallow(<NavigationBar
      setThemeKey={setTheme}
      currentThemeKey={'dark'}
      graph={graphMock}
      executionContext={'root'}
      scope={'root'}
    />);
    const button = element.find(Button);

    button.simulate('click');
    expect(setTheme).toBeCalledWith('light');
  });
});
