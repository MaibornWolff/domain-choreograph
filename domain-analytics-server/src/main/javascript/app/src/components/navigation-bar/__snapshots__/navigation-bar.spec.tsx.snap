// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationBar should render with graph and scope 1`] = `
<styled.nav>
  <styled.div>
    <styled.ul>
      <li
        key="contextApplication"
      >
        <styled.span>
          Application
        </styled.span>
      </li>
    </styled.ul>
  </styled.div>
  <styled.div />
</styled.nav>
`;

exports[`NavigationBar should render with graph and without scope 1`] = `
<styled.nav>
  <styled.div>
    <styled.ul>
      <li
        key="contextApplication"
      >
        <styled.span>
          Application
        </styled.span>
      </li>
    </styled.ul>
  </styled.div>
  <styled.div />
</styled.nav>
`;

exports[`NavigationBar should render without graph 1`] = `""`;
